src/jdk.jdi/share/classes/com/sun/tools/jdi/VMState.java

```diff
@@ -169,8 +170,32 @@ private synchronized void processVMAction(VMAction action) {
         }
     }
 
+    private final ReferenceQueue<VMListener> listenersReferenceQueue = new ReferenceQueue<>();
+
+    private void removeUnreachableListeners() {
+        // If there are no listeners on the ReferenceQueue, then that means none
+        // are unreachable and we can just return.
+        if (listenersReferenceQueue.poll() == null) {
+            return; // There are no unreachable listeners
+        }
+
+        // We always need to clear the ReferenceQueue
+        while (listenersReferenceQueue.poll() != null)
+            ;
```

Comment:
I'm not sure how does this works.
Do we keep pooling if there is a reference object in the queue?
Can it be an endless loop?

Reply:
The poll() removes the next object from the queue.

Reply:
This code is a bit odd in that it uses the reference queue to find out if there are one or more unreachable listeners. If there are, then it uses the code below to remove them. So the reference queue is really just a unreachable object detection mechanism, and not used for the actual of processing the objects.

Reply:
Okay. Thank you for explanation. It looks reasonable then.